### PR TITLE
docs(multisig-prover): write documentation on how WorkerSets are updated and confirmed

### PR DESCRIPTION
**Pull request by fahimahmedx**
## Description
[AXE-2606]
Updates the docs to include how worker set is automatically updated.

[AXE-2606]: https://axelarnetwork.atlassian.net/browse/AXE-2606?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: doc/src/contracts/multisig_prover.md [hunk 1]
```diff
@@ -62,17 +63,101 @@ Prover-->>-Relayer: returns GetProofResponse
...
+```
+
+1. Once the destination gateway emits a `OperatorshipTransferred` picked up by the Relayer, the Relayer calls Voting Verifier to create a poll. 
+2. Default Response is returned.
+3. Once the poll is completed, the Relayer calls the Prover to confirm if the `WorkerSet` was updated.
+4. The Prover queries the Voting Verifier to check if the `WorkerSet` is confirmed.
+5. The Voting Verifier returns if the `WorkerSet` is confirmed. If true, the `Prover` stores the `WorkerSet`.
+6. The new `WorkerSet` is stored in Multisig.
+7. Default response is returned.
+8. AppResponse is returned.
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:160]: Can we just have one section combining these two, with a sequence diagram as well as a graph that shows all of the components, including the relayer and the gateway on the external chain?

**Review by cjcobb23 — CHANGES_REQUESTED** [review]

File: doc/src/contracts/multisig_prover.md (hunk 1 above)
Comment (fahimahmedx) [reply doc/src/contracts/multisig_prover.md:160]: Yes, will do!

File: doc/src/contracts/multisig_prover.md [hunk 2]
```diff
@@ -62,17 +63,97 @@ Prover-->>-Relayer: returns GetProofResponse
...
+22. The Voting Verifier emits event `PollEnded`.
+23. Once the poll is completed, the Relayer calls the Prover to confirm if the `WorkerSet` was updated.
+24. The Prover queries the Voting Verifier to check if the `WorkerSet` is confirmed.
+25. The Voting Verifier returns if the `WorkerSet` is confirmed. If true, the `Prover` stores the `WorkerSet`.
+26. The new `WorkerSet` is stored in Multisig.
+27. Default response is returned.
+28. Response is returned.
+
 
 ## Interface
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:158]: Can we move this above the diagrams?

File: doc/src/contracts/multisig_prover.md [hunk 3]
```diff
@@ -62,17 +63,97 @@ Prover-->>-Relayer: returns GetProofResponse
...
+20. Default Response is returned.
+21. The Relayer calls the Voting Verifier to end the poll.
+22. The Voting Verifier emits event `PollEnded`.
+23. Once the poll is completed, the Relayer calls the Prover to confirm if the `WorkerSet` was updated.
+24. The Prover queries the Voting Verifier to check if the `WorkerSet` is confirmed.
+25. The Voting Verifier returns if the `WorkerSet` is confirmed. If true, the `Prover` stores the `WorkerSet`.
+26. The new `WorkerSet` is stored in Multisig.
+27. Default response is returned.
+28. Response is returned.
+
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:157]: still need a high level diagram, like the first diagram on this page.

Comment (fahimahmedx) [reply doc/src/contracts/multisig_prover.md:157]: yes! working on it as we speak :D

File: doc/src/contracts/multisig_prover.md (hunk 2 above)
Comment (fahimahmedx) [reply doc/src/contracts/multisig_prover.md:158]: Is it standard to have the steps above the diagram it's referencing? I thought it looked nicer keeping it below its diagram

File: doc/src/contracts/multisig_prover.md [hunk 4]
```diff
@@ -62,59 +109,122 @@ Prover-->>-Relayer: returns GetProofResponse
...
+r--UpdateWorkerSet-->b
+b--RegisterWorkerSet-->m
+b--StartSigningSession-->m
+r--GetProof-->b
+b--GetSigningSession-->m
+s--SubmitSignature-->m
+r-.->e
+e-.->r
+r--ConfirmWorkerSet-->v
+r--EndPoll-->v
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:149]: You can remove this too, it's not really necessary to understand the flow.

File: doc/src/contracts/multisig_prover.md [hunk 5]
```diff
@@ -62,59 +112,112 @@ Prover-->>-Relayer: returns GetProofResponse
...
+actor Signers
+Relayer->>+Prover: ExecuteMsg::UpdateWorkerSet
+alt no WorkerSet stored
+  Prover->>Prover: save new WorkerSet
+  Prover->>+Multisig: ExecuteMsg::RegisterWorkerSet
+  Multisig-->>-Prover: returns Response
+  Prover-->>-Relayer: returns Response
+else existing WorkerSet stored
+  Prover->>Prover: save new WorkerSet as the next WorkerSet
+  Prover->>+Multisig: ExecuteMsg::StartSigningSession
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:169]:
```suggestion
  Prover->>+Multisig: ExecuteMsg::StartSigningSession (for operatorship transferred message)
```

File: doc/src/contracts/multisig_prover.md [hunk 6]
```diff
@@ -62,59 +112,112 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Prover-->>Relayer: emit ProofUnderConstruction event
+  loop Collect signatures
+	  Signers->>+Multisig: signature collection
+  end
+  Multisig-->>-Relayer: emit SigningCompleted event
+  Relayer->>+Prover: QueryMsg::GetProof
+  Prover->>+Multisig: QueryMsg::GetSigningSession
+  Multisig-->>-Prover: reply with status, current signatures vector and snapshot
+  Prover-->>-Relayer: returns GetProofResponse
+  Relayer-->>External Chain: send proof and data
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:181]:
```suggestion
  Relayer-->>External Chain: send new workerset to the gateway, signed by old workerset
```

File: doc/src/contracts/multisig_prover.md [hunk 7]
```diff
@@ -62,59 +112,112 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  External Chain-->>+Relayer: emit OperatorshipTransferred event
+  Relayer->>+Voting Verifier: ExecuteMsg::ConfirmWorkerSet
+  Voting Verifier-->>-Relayer: returns Response
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:184]: these `returns Response` lines are not necessary and can be omitted. Makes the diagram easier to read, and they don't really add much

File: doc/src/contracts/multisig_prover.md [hunk 8]
```diff
@@ -62,59 +112,112 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Relayer->>+Voting Verifier: ExecuteMsg::EndPoll
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:185]: Probably want to show workers looking up the event on the external chain, and then voting.

File: doc/src/contracts/multisig_prover.md [hunk 9]
```diff
@@ -62,59 +112,112 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Voting Verifier-->>-Relayer: emit PollEnded event
+  Relayer->>+Prover: ExecuteMsg::ConfirmWorkerSet
+  Prover->>+Voting Verifier: QueryMsg::IsWorkerSetConfirmed
+  Voting Verifier-->>-Prover: query result
+  Prover->>+Multisig: ExecuteMsg::RegisterWorkerSet
+  Multisig-->>-Prover: returns Response
+  Prover-->>-Relayer: returns Response
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:192]: Don't need these lines

File: doc/src/contracts/multisig_prover.md (hunk 8 above)
Comment (fahimahmedx) [reply doc/src/contracts/multisig_prover.md:189]: Updated to include voting. I wasn't sure how the `handler` in `ampd`'s verify worker set was called, does the update look right?

File: doc/src/contracts/multisig_prover.md [hunk 10]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Prover-->>Relayer: emit ProofUnderConstruction event
+  loop Collect signatures
+	  Signers->>+Multisig: signature collection
+  end
+  Multisig-->>-Relayer: emit SigningCompleted event
+  Relayer->>+Prover: QueryMsg::GetProof
+  Prover->>+Multisig: QueryMsg::GetSigningSession
+  Multisig-->>-Prover: reply with status, current signatures vector and snapshot
+  Prover-->>-Relayer: returns GetProofResponse
+  Relayer-->>External Chain: send new WorkerSet to the gateway, signed by old WorkerSet
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:184]:
```suggestion
  Relayer-->>External Gateway: send new WorkerSet to the gateway, signed by old WorkerSet
```

File: doc/src/contracts/multisig_prover.md [hunk 11]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  External Chain-->>+Relayer: emit OperatorshipTransferred event
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:185]:
```suggestion
  External Gateway-->>+Relayer: emit OperatorshipTransferred event
```

File: doc/src/contracts/multisig_prover.md (hunk 8 above)
Comment (cjcobb23) [reply doc/src/contracts/multisig_prover.md:189]: When the poll starts, it triggers ampd to look up the event on the external gateway

File: doc/src/contracts/multisig_prover.md [hunk 12]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Relayer->>+Voting Verifier: ExecuteMsg::ConfirmWorkerSet
+  Voting Verifier-->>-Relayer: emit PollStarted event
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:187]: This line can be omitted, the relayer doesn't do anything with this event. Ampd sees this event, which triggers voting.

File: doc/src/contracts/multisig_prover.md [hunk 13]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Relayer->>+Voting Verifier: ExecuteMsg::Vote
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:188]:
```suggestion
  Worker->>+External Gateway: lookup OperatorshipTransferred event, verify event matches worker set in poll
  Worker->>+Voting Verifier: ExecuteMsg::Vote
```
the relayer is not voting. workers are voting

File: doc/src/contracts/multisig_prover.md [hunk 14]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Relayer->>+Voting Verifier: ExecuteMsg::EndPoll
+  Voting Verifier-->>-Relayer: emit PollEnded event
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:190]: this can be omitted. The relayer doesn't care about the PollEnded event, since it ended the poll itself.

File: doc/src/contracts/multisig_prover.md [hunk 15]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Relayer->>+Prover: ExecuteMsg::ConfirmWorkerSet
+  Prover->>+Voting Verifier: QueryMsg::IsWorkerSetConfirmed
+  Voting Verifier-->>-Prover: query result
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:193]:
```suggestion
  Voting Verifier-->>-Prover: true
```

File: doc/src/contracts/multisig_prover.md [hunk 16]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  Prover->>+Service Registry: QueryMsg::GetActiveWorkers
+  Service Registry-->>-Prover: save new WorkerSet as current WorkerSet
+  Prover->>+Multisig: ExecuteMsg::RegisterWorkerSet
+else existing WorkerSet stored
+  Prover->>+Service Registry: QueryMsg::GetActiveWorkers
+  Service Registry-->>-Prover: save new WorkerSet as next WorkerSet
+  Prover->>+Multisig: ExecuteMsg::StartSigningSession (for operatorship transferred message)
+  Multisig-->>Signers: emit SigningStarted event
+  Multisig->>-Prover: reply with session ID
+  Prover-->>Relayer: emit ProofUnderConstruction event
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:175]: You can omit these 3 lines, they are described in the previous diagram. This flow is more about the whole worker set update process. The nuances of signing proofs is best left to the previous diagram, to keep this one readable.

File: doc/src/contracts/multisig_prover.md [hunk 17]
```diff
@@ -62,59 +112,113 @@ Prover-->>-Relayer: returns GetProofResponse
...
+  loop Collect signatures
+	  Signers->>+Multisig: signature collection
+  end
+  Multisig-->>-Relayer: emit SigningCompleted event
+  Relayer->>+Prover: QueryMsg::GetProof
+  Prover->>+Multisig: QueryMsg::GetSigningSession
+  Multisig-->>-Prover: reply with status, current signatures vector and snapshot
+  Prover-->>-Relayer: returns GetProofResponse
```
Comment (cjcobb23) [inline doc/src/contracts/multisig_prover.md:183]: To keep things more readable, I'd shorten this to:
```suggestion
  Relayer->>+Prover: QueryMsg::GetProof
  Prover-->>-Relayer: returns GetProofResponse (new worker set signed by old worker set)
```

I know we are skipping a few steps, but we describe those steps in detail in the previous sequence diagram
